torch/_inductor/runtime/benchmarking.py

```diff
@@ -282,6 +352,129 @@ def benchmark_gpu(
+        grouping of CUDA event pairs.
+        """
+        return [
+            self.get_event_pairs_min_timing(list(event_pairs))
+            for event_pairs in zip(*interleaved_event_pairs)
+        ]
+
+    @time_and_count
+    def benchmark_many_gpu(
+        self: Self,
+        callables: List[Callable[[], Any]],
+        estimation_iters: int = 5,
+        memory_warmup_iters: int = 100,
+        benchmark_iters: int = 100,
+        max_benchmark_duration: int = 25,
+        **kwargs: Any,
+    ) -> List[float]:
+        """Benchmark many GPU callables using a custom benchmarking implementation.
+
+        Arguments:
+        - callables: The callables to benchmark.
+
+        Keyword Arguments:
+        - estimation_iters: The number of iterations to run `_callable` during
+        runtime estimation.
+        - memory_warmup_iters: The number of iterations to flush the L2 cache
+        before benchmarking.
+        - benchmark_iters: The number of iterations to run `_callable` during
+        benchmarking.
+        - max_benchmark_duration: The maximum duration of the benchmarking per
+        callable, in milliseconds. An estimated duration is calculated based on
+        the values of `memory_warmup_iters` and `benchmark_iters`, along with the
+        estimated runtime of `_callable` and various other factors, and we then
+        shrink `benchmark_iters` to fit in the alloted maximum duration.
+        - **kwargs: Additional kwargs that may be passed to the fallback.
+
+        Returns:
+        - The minimum runtime of each callable in `callables`, in milliseconds.
+        """
+        # we don't want any outside errors propagating into benchmarking
+        torch.cuda.synchronize()
+
+        # warmup each callable in `callables` (and catches any failures in the process)
+        for _callable in callables:
+            _callable()
+        torch.cuda.synchronize()
+
+        # see https://github.com/triton-lang/triton/pull/840 for why `dtype=torch.int`
+        buffer = torch.empty(self.L2_cache_size // 4, dtype=torch.int, device="cuda")
+        buffer.zero_()
+
+        # estimate the runtime of `_callable`
+        interleaved_event_pairs = self.get_interleaved_event_pairs(
+            len(callables), estimation_iters
+        )
+        for event_pairs in interleaved_event_pairs:
+            for _callable, (start_event, end_event) in zip(callables, event_pairs):
+                buffer.zero_()
+                start_event.record()
+                _callable()
+                end_event.record()
+        torch.cuda.synchronize()
+        estimated_timings = self.get_interleaved_event_pairs_min_timing(
+            interleaved_event_pairs
+        )
+
+        # adjust `benchmark_iters` to fit in the maximum benchmarking duration, we're
+        # alloted `max_benchmark_duration` per-callable, so we can just take the average
+        # of the estimated timings
+        benchmark_iters = max(
```

Comment:
I guess this runs the risk of very small fn + very large fn running the small fn not very often. should still be fine..